// @apify/storage-local/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApifyStorageLocal = void 0;
const tslib_1 = require("tslib");
const fs_extra_1 = require("fs-extra");
const ow_1 = tslib_1.__importDefault(require("ow"));
const path_1 = require("path");
const log_1 = tslib_1.__importDefault(require("@apify/log"));
const consts_1 = require("@apify/consts");
const database_connection_cache_1 = require("./database_connection_cache");
const dataset_1 = require("./resource_clients/dataset");
const dataset_collection_1 = require("./resource_clients/dataset_collection");
const key_value_store_1 = require("./resource_clients/key_value_store");
const key_value_store_collection_1 = require("./resource_clients/key_value_store_collection");
const request_queue_1 = require("./resource_clients/request_queue");
const request_queue_collection_1 = require("./resource_clients/request_queue_collection");
// Singleton cache to be shared across all ApifyStorageLocal instances
// to make sure that multiple connections are not created to the same database.
const databaseConnectionCache = new database_connection_cache_1.DatabaseConnectionCache();
/**
 * Represents local emulation of [Apify Storage](https://apify.com/storage).
 */
class ApifyStorageLocal {
    constructor(options = {}) {
        Object.defineProperty(this, "storageDir", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "requestQueueDir", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "keyValueStoreDir", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "datasetDir", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "dbConnections", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: databaseConnectionCache
        });
        Object.defineProperty(this, "enableWalMode", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * DatasetClient keeps internal state: itemCount
         * We need to keep a single client instance not to
         * have different numbers across parallel clients.
         */
        Object.defineProperty(this, "datasetClientCache", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new Map()
        });
        // To prevent directories from being created immediately when
        // an ApifyClient instance is constructed, we create them lazily.
        Object.defineProperty(this, "isRequestQueueDirInitialized", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "isKeyValueStoreDirInitialized", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "isDatasetDirInitialized", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        ow_1.default(options, 'ApifyStorageLocalOptions', ow_1.default.optional.object.exactShape({
            storageDir: ow_1.default.optional.string,
            enableWalMode: ow_1.default.optional.boolean,
        }));
        const { storageDir = './apify_storage', enableWalMode = true, } = options;
        this.storageDir = storageDir;
        this.requestQueueDir = path_1.resolve(storageDir, "request_queues" /* REQUEST_QUEUES */);
        this.keyValueStoreDir = path_1.resolve(storageDir, "key_value_stores" /* KEY_VALUE_STORES */);
        this.datasetDir = path_1.resolve(storageDir, "datasets" /* DATASETS */);
        this.enableWalMode = enableWalMode;
        this.dbConnections.setWalMode(this.enableWalMode);
    }
    datasets() {
        this._ensureDatasetDir();
        return new dataset_collection_1.DatasetCollectionClient({
            storageDir: this.datasetDir,
        });
    }
    dataset(id) {
        ow_1.default(id, ow_1.default.string);
        this._ensureDatasetDir();
        let client = this.datasetClientCache.get(id);
        if (!client) {
            client = new dataset_1.DatasetClient({
                name: id,
                storageDir: this.datasetDir,
            });
            this.datasetClientCache.set(id, client);
        }
        return client;
    }
    keyValueStores() {
        this._ensureKeyValueStoreDir();
        return new key_value_store_collection_1.KeyValueStoreCollectionClient({
            storageDir: this.keyValueStoreDir,
        });
    }
    keyValueStore(id) {
        ow_1.default(id, ow_1.default.string);
        this._ensureKeyValueStoreDir();
        return new key_value_store_1.KeyValueStoreClient({
            name: id,
            storageDir: this.keyValueStoreDir,
        });
    }
    requestQueues() {
        this._ensureRequestQueueDir();
        return new request_queue_collection_1.RequestQueueCollectionClient({
            storageDir: this.requestQueueDir,
            dbConnections: this.dbConnections,
        });
    }
    requestQueue(id, options = {}) {
        ow_1.default(id, ow_1.default.string);
        // Matching the Client validation.
        ow_1.default(options, ow_1.default.object.exactShape({
            clientKey: ow_1.default.optional.string,
        }));
        this._ensureRequestQueueDir();
        return new request_queue_1.RequestQueueClient({
            name: id,
            storageDir: this.requestQueueDir,
            dbConnections: this.dbConnections,
        });
    }
    _ensureDatasetDir() {
        if (!this.isDatasetDirInitialized) {
            fs_extra_1.ensureDirSync(this.datasetDir);
            this._checkIfStorageIsEmpty("Dataset" /* DATASET */, this.datasetDir);
            this.isDatasetDirInitialized = true;
        }
    }
    _ensureKeyValueStoreDir() {
        if (!this.isKeyValueStoreDirInitialized) {
            fs_extra_1.ensureDirSync(this.keyValueStoreDir);
            this._checkIfStorageIsEmpty("Key-value store" /* KEY_VALUE_STORE */, this.keyValueStoreDir);
            this.isKeyValueStoreDirInitialized = true;
        }
    }
    _ensureRequestQueueDir() {
        if (!this.isRequestQueueDirInitialized) {
            fs_extra_1.ensureDirSync(this.requestQueueDir);
            this._checkIfStorageIsEmpty("Request queue" /* REQUEST_QUEUE */, this.requestQueueDir);
            this.isRequestQueueDirInitialized = true;
        }
    }
    _checkIfStorageIsEmpty(storageType, storageDir) {
        const dirsWithPreviousState = [];
        const dirents = fs_extra_1.readdirSync(storageDir, { withFileTypes: true });
        for (const dirent of dirents) {
            if (!dirent.isDirectory())
                continue; // eslint-disable-line
            const innerStorageDir = path_1.resolve(storageDir, dirent.name);
            let innerDirents = fs_extra_1.readdirSync(innerStorageDir).filter((fileName) => !(/(^|\/)\.[^/.]/g).test(fileName));
            if (storageType === "Key-value store" /* KEY_VALUE_STORE */) {
                innerDirents = innerDirents.filter((fileName) => !RegExp(consts_1.KEY_VALUE_STORE_KEYS.INPUT).test(fileName));
            }
            if (innerDirents.length) {
                dirsWithPreviousState.push(innerStorageDir);
            }
        }
        const dirsNo = dirsWithPreviousState.length;
        if (dirsNo) {
            log_1.default.warning(`The following ${storageType} director${dirsNo === 1 ? 'y' : 'ies'} contain${dirsNo === 1 ? 's' : ''} a previous state:`
                + `\n      ${dirsWithPreviousState.join('\n      ')}`
                + '\n      If you did not intend to persist the state - '
                + `please clear the respective director${dirsNo === 1 ? 'y' : 'ies'} and re-start the actor.`);
        }
    }
}
exports.ApifyStorageLocal = ApifyStorageLocal;
;
