// apify-client/dist/resource_clients/actor_env_var_collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorEnvVarCollectionClient = void 0;
const tslib_1 = require("tslib");
const ow_1 = (0, tslib_1.__importDefault)(require("ow"));
const resource_collection_client_1 = require("../base/resource_collection_client");
/**
 * @hideconstructor
 */
class ActorEnvVarCollectionClient extends resource_collection_client_1.ResourceCollectionClient {
    constructor(options) {
        super({
            resourcePath: 'env-vars',
            ...options,
        });
    }
    /**
     * TODO: https://docs.apify.com/api/v2#/reference/actors/env-var-collection/get-list-of-env-vars
     */
    async list(options = {}) {
        (0, ow_1.default)(options, ow_1.default.object.exactShape({
            limit: ow_1.default.optional.number,
            offset: ow_1.default.optional.number,
            desc: ow_1.default.optional.boolean,
        }));
        return this._list(options);
    }
    /**
     * TODO: https://docs.apify.com/api/v2#/reference/actors/env-var-collection/create-env-var
     */
    async create(actorEnvVar) {
        (0, ow_1.default)(actorEnvVar, ow_1.default.optional.object);
        return this._create(actorEnvVar);
    }
}
exports.ActorEnvVarCollectionClient = ActorEnvVarCollectionClient;
