// @apify/log/log_consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LEVEL_TO_STRING = exports.LEVELS = exports.PREFIX_DELIMITER = exports.LogLevel = void 0;
var LogLevel;
(function (LogLevel) {
    // Turns off logging completely
    LogLevel[LogLevel["OFF"] = 0] = "OFF";
    // For unexpected errors in Apify system
    LogLevel[LogLevel["ERROR"] = 1] = "ERROR";
    // For situations where error is caused by user (e.g. Meteor.Error), i.e. when the error is not
    // caused by Apify system, avoid the word "ERROR" to simplify searching in log
    LogLevel[LogLevel["SOFT_FAIL"] = 2] = "SOFT_FAIL";
    LogLevel[LogLevel["WARNING"] = 3] = "WARNING";
    LogLevel[LogLevel["INFO"] = 4] = "INFO";
    LogLevel[LogLevel["DEBUG"] = 5] = "DEBUG";
    // for performance stats
    LogLevel[LogLevel["PERF"] = 6] = "PERF";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
exports.PREFIX_DELIMITER = ':';
exports.LEVELS = LogLevel;
// Inverse of LOG_LEVELS = maps log level to string.
exports.LEVEL_TO_STRING = Object.keys(LogLevel).filter((x) => Number.isNaN(+x));
